test/e2e/framework/kubectl/kubectl_util.go

```diff
@@ -0,0 +1,292 @@
+
+var (
+	// For parsing Kubectl version for version-skewed testing.
+	gitVersionRegexp = regexp.MustCompile("GitVersion:\"(v.+?)\"")
+)
+
+// KubectlBuilder is used to build, customize and execute a kubectl Command.
+// Add more functions to customize the builder as needed.
+type KubectlBuilder struct {
+	// Cmd is the custom command executed by KubectlBuilder.
+	Cmd     *exec.Cmd
+	timeout <-chan time.Time
+}
+
+// TestContext is a struct containing the minimum attributes needed to run KubectlCmd.
+type TestContext struct {
+	CertDir     string
+	Host        string
+	KubeConfig  string
+	KubeContext string
+	KubectlPath string
+}
+
+// NewTestContext returns a new Kubeconfig struct instance.
+func NewTestContext(certdir string, host string, kubeconfig string, kubecontext string, kubectlpath string) *TestContext {
+	return &TestContext{
+		CertDir:     certdir,
+		Host:        host,
+		KubeConfig:  kubeconfig,
+		KubeContext: kubecontext,
+		KubectlPath: kubectlpath,
+	}
+}
+
+// KubectlCmd runs the kubectl executable through the wrapper script.
+func (tc *TestContext) KubectlCmd(args ...string) *exec.Cmd {
+	defaultArgs := []string{}
+
+	// Reference a --server option so tests can run anywhere.
+	if tc.Host != "" {
+		defaultArgs = append(defaultArgs, "--"+clientcmd.FlagAPIServer+"="+tc.Host)
+	}
+	if tc.KubeConfig != "" {
+		defaultArgs = append(defaultArgs, "--"+clientcmd.RecommendedConfigPathFlag+"="+tc.KubeConfig)
+
+		// Reference the KubeContext
+		if tc.KubeContext != "" {
+			defaultArgs = append(defaultArgs, "--"+clientcmd.FlagContext+"="+tc.KubeContext)
+		}
+
+	} else {
+		if tc.CertDir != "" {
+			defaultArgs = append(defaultArgs,
+				fmt.Sprintf("--certificate-authority=%s", filepath.Join(tc.CertDir, "ca.crt")),
+				fmt.Sprintf("--client-certificate=%s", filepath.Join(tc.CertDir, "kubecfg.crt")),
+				fmt.Sprintf("--client-key=%s", filepath.Join(tc.CertDir, "kubecfg.key")))
+		}
+	}
+	kubectlArgs := append(defaultArgs, args...)
+
+	//We allow users to specify path to kubectl, so you can test either "kubectl" or "cluster/kubectl.sh"
+	//and so on.
+	cmd := exec.Command(tc.KubectlPath, kubectlArgs...)
+
+	//caller will invoke this and wait on it.
+	return cmd
+}
+
+// KubectlVersion gets the version of kubectl that's currently being used (see
+// --kubectl-path in e2e.go to use an alternate kubectl).
+func KubectlVersion() (*utilversion.Version, error) {
+	args := []string{"version", "--client"}
+	output := RunKubectlOrDie(args...)
+	matches := gitVersionRegexp.FindStringSubmatch(output)
+	if len(matches) != 2 {
+		return nil, fmt.Errorf("Could not find kubectl version in output %v", output)
+	}
+	// Don't use the full match, as it contains "GitVersion:\"" and a
+	// trailing "\"".  Just use the submatch.
+	return utilversion.ParseSemantic(matches[1])
+}
+
+// NewKubectlCommand returns a KubectlBuilder for running kubectl.
+func NewKubectlCommand(tc TestContext, args ...string) *KubectlBuilder {
```

Comment:
This is where I am stuck..

Should I:
* Option 1: Add a parameter to the function so that it takes in a `TestContext` so that we can call `tc.KubectlCmd`.
* Option 2: Make `NewKubectlCommand` a method on the receiver like: `func (tc *TestContext) NewKubectlCommand(args ...string) *KubectlBuilder`

This (`NewKubectlCommand`)is used in the bottom 4 functions in this file (`RunKubectl`, `RunKubectlOrDie`, `RunKubectlInput `, `RunKubectlOrDieInput`). And those 4 functions are used in numerous places throughout the codebase..

I've been trying to figure out the best, simplest way to do this. I may be overthinking this... but this is where I'm at currently.

Reply:
I'm in favor of option 1.
This way the the `KubectlCmd()` functionality will be independent of the `KubeBuilder`.
The former can and is used throughout the place to run one off commands, i.e.,

```
test/e2e/framework/framework.go:681:	cmd := KubectlCmd(cmdArgs...)
test/e2e/framework/util.go:1232:// KubectlCmd runs the kubectl executable through the wrapper script.
test/e2e/framework/util.go:1233:func KubectlCmd(args ...string) *exec.Cmd {
test/e2e/framework/util.go:1303:	b.cmd = KubectlCmd(args...)
test/e2e/windows/memory_limits.go:266:	cmd := framework.KubectlCmd("proxy", "-p", "0")
test/e2e/node/pods.go:106:			cmd := framework.KubectlCmd("proxy", "-p", "0")
test/e2e/kubectl/kubectl.go:1860:			cmd := framework.KubectlCmd("proxy", fmt.Sprintf("--unix-socket=%s", path))
test/e2e/kubectl/kubectl.go:2116:	cmd := framework.KubectlCmd("proxy", "-p", "0", "--disable-filter")
test/e2e/kubectl/portforward.go:171: 	cmd := framework.KubectlCmd("port-forward", fmt.Sprintf("--namespace=%v", ns), podName, fmt.Sprintf(":%d", port))
```

Reply:
ok, thank you for your feedback, will update